### PR TITLE
[Swift] Enum parameters are now handled

### LINKED ISSUE #2531: [Swift] enum parameters do not generate enums
```js         
               "name": "paramName"
               "required": false,
               "type": "string",
               "enum": [
                   "one",
                   "two"
               ]
```
generates:

```swift
    public class func action(paramName: String?)
```

while ideally it should generates:

```swift
   enum ParamName: String {
      case One = "one"
      case Two = "two"
   }
   public class func action(paramName: ParamName?)
```

### PR DESCRIPTION
**Pull request by fabdslv**
Fix #2531

### CODE REVIEW

**Comment by yonaskolb** [comment]
This won't compile if multiple operations have the same enum. Maybe you could prepend the operationId to the enum type?

**Comment by wing328** [comment]
@fabdslv thanks for the PR.

@yonaskolb thanks for reviewing.

I've a PR for C# and PHP (https://github.com/swagger-api/swagger-codegen/pull/2508) and should work for multiple operations with the same enum defined.

I'll try to incorporate your change into my PR.

**Comment by yonaskolb** [comment]
Well it will break in swift if there are multiple operations with the same tag that share any enums with the same name, as they are in the same scope

**Comment by yonaskolb** [comment]
The rawValue of the enum will also need to be extracted when added to the path parameters or the nillableParameters, otherwise they will use the printed escaped enum name which may be different from the actual value

**Comment by fabdslv** [comment]
@yonaskolb Yes, it might be a good solution to add the operational id. I'm not sure I understand your last comment though, what do you mean by "extract"?

**Comment by yonaskolb** [comment]
well for the path params instead of 
`"\({{paramName}})"`
use
`"\({{paramName}}{{#isEnum}}.rawValue{{/isEnum}})"`

and for the query params instead of
`"{{baseName}}": {{paramName}}` 
use 
`"{{baseName}}": {{paramName}}}{{#isEnum}}{{^required}}?{{/required}}.rawValue{{/isEnum}}`

I hope that makes sense. That's what I've done in my custom templates

**Comment by wing328** [comment]
@yonaskolb when you've time, do you mind submitting your custom template change with a PR so that we can review and compare? Thanks!

**Comment by fabdslv** [comment]
I updated the PR to integrate what @yonaskolb was mentioning. Apart from the fact that it fixes the issues, now the Enum name is ```{{{datatypeWithEnum}}}_{{operationId}}``` so that it fixes the conflicts we might encounter.

@wing328 It seems that some changes were made in master that affect my commit and I cannot rebase from master. Could you please take a look at this if you have time? Thanks.

File: modules/swagger-codegen/src/main/resources/swift/api.mustache [hunk 1]
```diff
@@ -16,14 +16,26 @@ extension {{projectName}}API {
 /** {{description}} */{{/description}}
 public class {{classname}}: APIBase {
 {{#operation}}
+    {{#allParams}}
+    {{#isEnum}}
+    /**
+
+    enum for parameter {{paramName}}
+    */
+    public enum {{{datatypeWithEnum}}}_{{operationId}}: String { {{#allowableValues}}{{#values}}
```
Comment (cjolif) [inline modules/swagger-codegen/src/main/resources/swift/api.mustache]: This is solving the possible clash issue. But for people that don't have a clash, this is created a cumbersome API. Maybe we should add some logic Java-side to know if we have a conflict and do that only in that case?

**Comment by fabdslv** [comment]
@wing328 I managed to rebase, it should be okay now.

File: modules/swagger-codegen/src/main/resources/swift/api.mustache (hunk 1 above)
Comment (wing328) [inline modules/swagger-codegen/src/main/resources/swift/api.mustache]: @cjolif I agree with your comment.

@fabdslv I've a pending PR (https://github.com/swagger-api/swagger-codegen/pull/2508) that should be able to handle the situation.

**Comment by wing328** [comment]
@fabdslv thanks! I'll merge this into https://github.com/swagger-api/swagger-codegen/pull/2508 so that we only have javascript and typescript to add better enum support. Hopefully that can be done this weekend...

**Pull request merged**